11-apply-manifest-code/README.md

```diff
@@ -0,0 +1,245 @@
+26c47f8c4bad        lab_node            "/usr/sbin/sshd -D"   About an hour ago   Up About an hour    0.0.0.0:20022->22/tcp, 0.0.0.0:20080->80/tcp   lb
+```
+
+# Build a Boltdir
+
+By default `$HOME/.puppetlabs/bolt/` is the base directory for user-supplied data such as the configuration and inventory files. It is effectively the default `Boltdir`. 
+You may find it useful to maintain a project specific `Boltdir`. When you commit a `Boltdir` to a project you can share Bolt configuration and code between users.
+
+Bolt will search for a `Boltdir` in parent directories of the directory from which it was run.
+
+## Inventory
+
+Build an inventory to organize provisioned nodes. This will be the first configuration file in our new project specific `Boltdir`. 
+
+```yaml
+---
+groups:
+  - name: lb
+    nodes:
+      - name: "0.0.0.0:20022"
+  - name: servers
+    nodes:
+      - name: "127.0.0.1"
+        config:
+          ssh:
+            port: 20023
+      - name: "localhost"
+        config:
+          ssh:
+            port: 20024
+config:
+  ssh:
+    host-key-check: false
+    password: root
+    user: root
+```
+Make sure your inventory is configured correctly and you can connect to all nodes. Run from within the project Boltdir:
+
+```bash
+bolt command run 'echo hi' -n all
+```
+Expected output
+```
+Started on 0.0.0.0...
+Started on localhost...
+Started on 127.0.0.1...
+Finished on localhost:
+  STDOUT:
+    hi
+Finished on 0.0.0.0:
+  STDOUT:
+    hi
+Finished on 127.0.0.1:
+  STDOUT:
+    hi
+Successful on 3 nodes: 0.0.0.0:20022,127.0.0.1,localhost
+Ran on 3 nodes in 0.20 seconds
+
+```
+
+## Module Content
+
+In order to instal module content from the forge Bolt uses a `Puppetfile`. See [Puppetfile Documentation](https://puppet.com/docs/pe/2017.3/puppetfile.html) for more information. 
+
+Save the following `Puppetfile` that describes the Puppet Forge content to be installed in the project `Boltdir`. 
+
+```
+forge 'http://forge.puppetlabs.com'
+mod 'puppetlabs-stdlib', '4.25.1'
+mod 'puppetlabs-concat', '4.2.1'
+mod 'puppetlabs-haproxy', '2.1.0'
+```
+
+> **Note**: In order to preserve local modules you must add them to the puppetfile. See the [Puppetfile](Boltdir/Puppetfile) in this repository for an example. 
+From within the `Boltdir` install the Forge content with the following Bolt command:
+```
+bolt puppetfile install
+```
+Confirm that a `modules` directory has been created in the project `Boltdir`. 
+
+## Write profile module
+
+Now that you have downloaded existing modules it is time to write your own module content. In the `Boltdir/modules` directory create a new directory called `profiles`. The `profiles` module will be our own custom module. 
+
+Start by abstracting the nginx setup by writing a Puppet Class. Puppet code belongs in a subdirectory of our module called `manifests`. Save the following class definition in `Boltdir/modules/profiles/manifests/server.pp`. 
```

Comment:
What would be a good link for "getting started with puppet" or something for those who have never seen a class declaration?

Reply:
Probably https://learn.puppet.com.

Reply:
The Learning VM is pretty good.